subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultFileCollectionSnapshotter.java

```diff
@@ -70,21 +68,20 @@ public void visitGenericFileTree(FileTreeInternal fileTree) {
 
         @Override
         public void visitFileTree(File root, PatternSet patterns, FileTreeInternal fileTree) {
-            MerkleDirectorySnapshotBuilder builder = MerkleDirectorySnapshotBuilder.sortingRequired();
-            virtualFileSystem.read(root.getAbsolutePath(), new PatternSetSnapshottingFilter(patterns, stat), builder);
-            FileSystemLocationSnapshot result = builder.getResult();
-            roots.add((result == null || result.getType() == FileType.Missing) ? FileSystemSnapshot.EMPTY : result);
+            virtualFileSystem.read(
+                root.getAbsolutePath(),
+                new PatternSetSnapshottingFilter(patterns, stat),
+                snapshot -> {
+                    if (snapshot.getType() != FileType.Missing) {
+                        roots.add(snapshot);
```

Comment:
Are the snapshots here still (potentially) mutable? If so, we should make a defensive copy.